

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Day 1 Afternoon &mdash; Micro 612 genomics workshop 3.0 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700&subset=latin,cyrillic' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Micro 612 genomics workshop 3.0 documentation" href="index.html"/>
        <link rel="next" title="Day 2 Morning" href="day2_morning.html"/>
        <link rel="prev" title="Day 1 Morning" href="day1_morning.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        
          <a href="index.html" class="fa fa-home"> Micro 612 genomics workshop</a>
        
        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="day1_morning.html">Day 1 Morning</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#installing-and-setting-up-cyberduck-for-file-transfer">Installing and setting up Cyberduck for file transfer</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#getting-your-data-onto-flux-and-setting-up-environment-variable">Getting your data onto Flux and setting up environment variable</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#unix-is-your-friend">Unix is your friend</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#your-first-sequence-analysis-program">Your first sequence analysis program!!!</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#plotting-genomic-coverage-in-r">Plotting genomic coverage in R</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#power-of-unix-commands">Power of Unix commands</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#contamination-screening-using-fastq-screen">Contamination Screening using FastQ Screen</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#quality-control-using-fastqc">Quality Control using FastQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#quality-trimming-using-trimmomatic">Quality Trimming using Trimmomatic</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Day 1 Afternoon</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#read-mapping">Read Mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generate-alignment-statistics">Generate Alignment Statistics</a></li>
<li class="toctree-l2"><a class="reference internal" href="#variant-calling-and-filteration">Variant Calling and Filteration</a></li>
<li class="toctree-l2"><a class="reference internal" href="#visualize-bam-and-vcf-files-in-artemis">Visualize BAM and VCF files in Artemis</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vre-variant-calling-analysis">VRE variant calling analysis</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="day2_morning.html">Day 2 Morning</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#genome-assembly-using-spades-pipeline">Genome Assembly using Spades Pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#assembly-evaluation-using-quast">Assembly evaluation using QUAST</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#generating-multiple-sample-reports-using-multiqc">Generating multiple sample reports using multiqc</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#compare-assembly-to-reference-genome-and-post-assembly-genome-improvement">Compare assembly to reference genome and post-assembly genome improvement</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#map-reads-to-the-final-ordered-assembly">Map reads to the final ordered assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#using-abacas-and-act-to-compare-vre-vse-genome">Using abacas and ACT to compare VRE/VSE genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#genome-annotation">Genome Annotation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="day2_afternoon.html">Day 2 Afternoon</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#high-throughput-blast-and-pan-genome-analysis">High-throughput BLAST and pan-genome analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#determine-which-genomes-contain-beta-lactamase-genes">Determine which genomes contain beta-lactamase genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#identification-of-antibiotic-resistance-genes-with-ariba-directly-from-paired-end-reads">Identification of antibiotic resistance genes with ARIBA directly from paired end reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#perform-pan-genome-analysis-with-roary">Perform pan-genome analysis with Roary</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#perform-genome-comparisons-with-act">Perform genome comparisons with ACT</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="day3_morning.html">Day 3 Morning</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#perform-whole-genome-alignment-with-mauve-and-convert-alignment-to-other-useful-formats">Perform whole genome alignment with Mauve and convert alignment to other useful formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#perform-some-dna-sequence-comparisons-and-phylogenetic-analysis-in-ape-an-r-package">Perform some DNA sequence comparisons and phylogenetic analysis in APE, an R package</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#perform-snp-density-analysis-to-discern-evidence-of-recombination">Perform SNP density analysis to discern evidence of recombination</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#perform-recombination-filtering-with-gubbins">Perform recombination filtering with Gubbins</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#create-annotated-publication-quality-trees-with-itol">Create annotated publication quality trees with iTOL</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="day3_afternoon.html">Day 3 Afternoon</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#klebsiella-pneumoniae-comparative-genomic-analysis">Klebsiella pneumoniae comparative genomic analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#perform-qc-on-fastq-files">Perform QC on fastq files</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#examine-results-of-spandx-pipeline">Examine results of SPANDx pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#recombination-detection-and-tree-generation">Recombination detection and tree generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#phylogenetic-tree-annotation-and-visualization">Phylogenetic tree annotation and visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#assessment-of-genomic-deletions">Assessment of genomic deletions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="online_resources.html">Helpful resources for microbial genomics</a></li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Micro 612 genomics workshop</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Day 1 Afternoon</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/day1_afternoon.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="day-1-afternoon">
<span id="day-1-afternoon"></span><h1>Day 1 Afternoon<a class="headerlink" href="#day-1-afternoon" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="index.html">[HOME]</a></p>
<p>Earlier this morning, We performed some quality control steps on our sequencing data to make it clean and usable for various downstream analysis. Now we will perform our first sequence analysis, specifically variant calling, and map these reads to a reference genome and try to find out the differences between them.</p>
<p>Read Mapping is one of the most common Bioinformatics operations that needs to be carried out on NGS data. The main goal behind read mapping/aligning is to find the best possible reference genome position to which reads could be aligned. Reads are generally mapped to a reference genome sequence that is sufficiently closely related genome to accurately align reads. There are number of tools that can map reads to a reference genome and they differ from each other in algorithm, speed and accuracy. Most of these tools work by first building an index of reference sequence which works like a dictionary for fast search/lookup and then applying an alignment algorithm that uses these index to align short read sequences against the reference.</p>
<p>These alignment has a vast number of uses, including:</p>
<ol class="simple">
<li>variant/SNP calling: Finding differences between your sequenced organism genome and the reference genome</li>
<li>coverage estimation: If you have sufficient reads to cover each position of reference genome.</li>
<li>gene expression analysis: determining the level of expression of each genes in a genome.</li>
</ol>
<p>In this session, we will be covering the important steps that are part of any Read mapping/Variant calling bioinformatics pipleine.</p>
<div class="section" id="read-mapping">
<span id="read-mapping"></span><h2>Read Mapping<a class="headerlink" href="#read-mapping" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="day1_afternoon.html">[back to top]</a>
<a class="reference external" href="index.html">[HOME]</a></p>
<p><img alt="alt tag" src="_images/1_1.png" /></p>
<p><strong>1. Navigate to your workshop home directory and copy day1_after directory from shared data directory.</strong></p>
<pre class="literal-block">
wd

cp -r /scratch/micro612w18_fluxod/shared/data/day1_after ./
</pre>
<p>We will be using trimmed clean reads that were obtained after running Trimmomatic on raw reads.</p>
<p><strong>2. Map your reads against a finished reference genome using <a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA</a></strong></p>
<p>Choosing the right read mapper is crucial and should be based on the type of analysis and data you are working with. Each aligners are meant to be better used with specific types of data, for example:</p>
<p>For whole genome or whole exome sequencing data: Use BWA for long reads (&gt; 50/100 bp), use Bowtie2 for short reads (&lt; 50/100bp)
For transcriptomic data (RNA-Seq): use Splice-aware Mapper such as Tophat. (Not applicable for microbial data)</p>
<p>Here, we will be using BWA aligner to map the reads against a reference genome, KPNIH1.</p>
<p>BWA is one of the several read mappers that are based on Burrows-Wheeler transform algorithm. If you feel like challenging yourselves, you can read BWA paper <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/25/14/1754.short">here</a></p>
<p>Read Mapping is a time-consuming step that involves searching the reference and finding the optimal location for the alignment for millions of reads. Creating an index file of a reference sequence for quick lookup/search operations significantly decreases the time required for read alignment. Imagine indexing a genome sequence like the index at the end of a book. If you want to know on which page a word appears or a chapter begins, it is much more efficient to look it up in a pre-built index than going through every page of the book. Similarly, an index of a large DNA sequence allows aligners to rapidly find shorter sequences embedded within it.</p>
<p>Note: each read mapper has its own unique way of indexing a reference genome and therefore the reference index created by BWA cannot be used for Bowtie. (Most Bioinformatics tools nowadays require some kind of indexing or reference database creation)</p>
<blockquote>
<div><strong><em>i. To create BWA index of Reference, you need to run following command.</em></strong></div></blockquote>
<p>Start a flux interactive session</p>
<pre class="literal-block">
iflux
</pre>
<p>Navigate to day1_after folder that you recently copied and create a new folder Rush_KPC_266_varcall_result for saving this exercise&#8217;s output.</p>
<pre class="literal-block">
d1a

#or 

cd /scratch/micro612w18_fluxod/username/day1_after/

mkdir Rush_KPC_266_varcall_result

</pre>
<p>Create bwa index for the reference genome.</p>
<pre class="literal-block">
bwa index KPNIH1.fasta
</pre>
<p>Also go ahead and create fai index file using samtools required by GATK in later downstream steps.</p>
<pre class="literal-block">
samtools faidx KPNIH1.fasta
</pre>
<blockquote>
<div><strong><em>ii. Align reads to reference and redirect the output into SAM file</em></strong></div></blockquote>
<p>Quoting BWA:
&#8220;BWA consists of three algorithms: BWA-backtrack, BWA-SW and BWA-MEM. The first algorithm is designed for Illumina sequence reads up to 100bp, while the rest two for longer sequences ranged from 70bp to 1Mbp. BWA-MEM and BWA-SW share similar features such as long-read support and split alignment, but BWA-MEM, which is the latest, is generally recommended for high-quality queries as it is faster and more accurate. BWA-MEM also has better performance than BWA-backtrack for 70-100bp Illumina reads.&#8221;</p>
<p>For other algorithms employed by BWA, you can refer to BWA <a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">manual</a></p>
<p>Now lets align both left and right end reads to our reference using BWA alignment algorithm &#8216;mem&#8217;.</p>
<pre class="literal-block">

bwa mem -M -R &quot;&#64;RG\tID:96\tSM:Rush_KPC_266_1_combine.fastq.gz\tLB:1\tPL:Illumina&quot; -t 8 KPNIH1.fasta forward_paired.fq.gz reverse_paired.fq.gz &gt; Rush_KPC_266_varcall_result/Rush_KPC_266__aln.sam

</pre>
<p>Read group tells aligners/other tools that certain reads were sequenced together on a specific lane. If you have multiplexed samples in a single lane, you will get multiple samples in a single read group. If you sequenced the same sample in several lanes, you will have multiple read groups for the same sample.</p>
<p>This string with -R flag says that all reads belongs to ID:96 and library LB:1; with sample name SM:Rush_KPC_266_1_combine.fastq.gz and was sequenced on illumina platform PL:Illumina.</p>
<p>You can extract this information from fastq read header. (&#64;M02127:96:000000000-AG04W:1:1101:13648:1481 1:N:0:44)</p>
<p><strong>3. SAM/BAM manipulation and variant calling using <a class="reference external" href="http://www.htslib.org/doc/samtools.html">Samtools</a></strong></p>
<blockquote>
<div><strong><em>i. Change directory to results folder and look for BWA output:</em></strong></div></blockquote>
<pre class="literal-block">
cd Rush_KPC_266_varcall_result

ls
</pre>
<p>The output of BWA and most of the short-reads aligners is a SAM file. SAM format is considered as the standard output for most read aligners and stands for Sequence Alignment/Map format. It is a TAB-delimited format that describes how each reads were aligned to the reference sequence.</p>
<p>Lets explore first few lines of .sam file.</p>
<pre class="literal-block">

head -n4 Rush_KPC_266__aln.sam

</pre>
<p>example:</p>
<pre class="literal-block">

&#64;SQ     SN:gi|661922017|gb|CP008827.1|  LN:5394056        &lt;=== Reference Genome name and its length
&#64;RG     ID:96   SM:Rush_KPC_266_1_combine.fastq.gz      LB:1    PL:Illumina &lt;=== sample read group info
&#64;PG     ID:bwa  PN:bwa  VN:0.7.12-r1039 CL:bwa mem -M -R &#64;RG\tID:96\tSM:Rush_KPC_266_1_combine.fastq.gz\tLB:1\tPL:Illumina -t 8 KPNIH1.fasta forward_paired.fq.gz reverse_paired.fq.gz       &lt;== aligner command 
M02127:96:000000000-AG04W:1:1101:23094:1725     99      gi|661922017|gb|CP008827.1|     4724728 60      250M    =       4724852 295     GCTGCCTGCAGCATCTCAGCGGCTTTATCGGCTCGCAGCAGGTGCGGCTGGTGACCCTCTCCGGCGGCGTCGGCCCGTATATGACCGGTATCGGCCAGCTTGATGCCGCCTGCAGCGTCAGCATTATCCCGGCGCCGCTGCGGGTCTCTTCGGCGGAGGTCTCCGAGATCCTGCGCCGCGAGTCGAGCGTGCGCGACGTGATCCTCGCGGCGACGGCGGCGGACGCGGCGGTAGTCGGCCTTGGCGCCAT      CCCCCGGGGGGGGGGEGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGG&#64;FGFGFFGGGGGGGGGGGGCFGBEGGGCFGGGFGDE&gt;*CGEFCCFCEECCCCGGGDGE5E&gt;5EEFEEGD=C=EDCE=EEECCC?C9CCECEDC&lt;DGGGGCDGG:CBC)&lt;DB&gt;&#64;EF??&gt;&gt;&#64;)7&lt;6?6354,4      NM:i:2  MD:Z:161G77A10  AS:i:240        XS:i:0  RG:Z:96

</pre>
<p>The lines starting with &#8220;&#64;&#8221; is a header section and contains information about reference genome, sample read group and the aligner command that was used for aligning the samples. The header section is followed by an alignment section information for each read. It contains 11 columns and an optional TAG option.</p>
<p>Detailed information about these 11 columns can be obtained from this <a class="reference external" href="https://www.google.com/url?sa=t&amp;rct=j&amp;q=&amp;esrc=s&amp;source=web&amp;cd=1&amp;ved=0ahUKEwizkvfAk9rLAhXrm4MKHVXxC9kQFggdMAA&amp;url=https%3A%2F%2Fsamtools.github.io%2Fhts-specs%2FSAMv1.pdf&amp;usg=AFQjCNHFmjxTXKnxYqN0WpIFjZNylwPm0Q">pdf</a> document.</p>
<p>The second column consists of coded bitwise flags where each code flag carries important information about the alignment. Open <a class="reference external" href="https://broadinstitute.github.io/picard/explain-flags.html">this</a> site and enter the flag &#8220;99&#8221; to find out what it stands for.</p>
<p>The last section &#8220;NM:i:2  MD:Z:161G77A10  AS:i:240 XS:i:0  RG:Z:96&#8221; is an optional tag section and varies for different aligners(specifications based on aligners).</p>
<p>Here,</p>
<p>NM tag tells number of changes necessary to make it equal to the reference(2 changes)</p>
<p>MD tag tells you what positions in the read alignment are different from reference base and is used by variant callers to call SNP&#8217;s. For example, The tag &#8220;MD:Z:161G77A10&#8221; implies that position 162 in the read carries a different base whereas the reference genome carries base &#8220;G&#8221;</p>
<p>AS is an alignment score and XS:i:0 is an suboptimal alignment score.</p>
<blockquote>
<div><strong><em>ii. Convert SAM to BAM using SAMTOOLS:</em></strong></div></blockquote>
<p>BAM is the compressed binary equivalent of SAM but are usually quite smaller in size than SAM format. Since, parsing through a SAM format is slow, Most of the downstream tools require SAM file to be converted to BAM so that it can be easily sorted and indexed.</p>
<p>The below command will ask samtools to convert SAM format(-S) to BAM format(-b)</p>
<pre class="literal-block">
samtools view -Sb Rush_KPC_266__aln.sam &gt; Rush_KPC_266__aln.bam
</pre>
<blockquote>
<div><strong><em>iii. Sort BAM file using SAMTOOLS:</em></strong></div></blockquote>
<p>Most of the downstream tools such as GATK requires your BAM file to be indexed and sorted by reference genome positions.</p>
<p>Now before indexing this BAM file, we will sort the data by positions(default) using samtools. Some RNA Seq/Gene expression tools require it to be sorted by read name which is achieved by passing -n flag.</p>
<pre class="literal-block">
samtools sort Rush_KPC_266__aln.bam Rush_KPC_266__aln_sort
</pre>
<p><strong>4. Mark duplicates(PCR optical duplicates) and remove them using <a class="reference external" href="http://broadinstitute.github.io/picard/command-line-overview.html#MarkDuplicates">PICARD</a></strong></p>
<p>Illumina sequencing involves PCR amplification of adapter ligated DNA fragments so that we have enough starting material for sequencing. Therefore, some amount of duplicates are inevitable. Ideally, you amplify upto ~65 fold(4% reads) but higher rates of PCR duplicates e.g. 30% arise when people have too little starting material such that greater amplification of the library is needed or some smaller fragments which are easier to PCR amplify, end up over-represented.</p>
<p>For an in-depth explanation about how PCR duplicates arise in sequencing, please refer to this interesting <a class="reference external" href="http://www.cureffi.org/2012/12/11/how-pcr-duplicates-arise-in-next-generation-sequencing/">blog</a></p>
<p>Picard identifies duplicates by searching reads that have same start position on reference or in PE reads same start for both ends. It will choose a representative from each group of duplicate reads based on best base quality scores and other criteria and retain it while removing other duplicates. This step plays a significant role in removing false positive variant calls(such as sequencing error) during variant calling that are represented by PCR duplicate reads.</p>
<p><img alt="alt tag" src="_images/picard.png" /></p>
<blockquote>
<div><strong><em>i. Create a dictionary for reference fasta file required by PICARD</em></strong></div></blockquote>
<p>Make sure you are in Rush_KPC_266_varcall_result directory and are giving proper reference genome path (day1_after directory).</p>
<pre class="literal-block">

java -jar /scratch/micro612w18_fluxod/shared/bin/picard-tools-1.130/picard.jar CreateSequenceDictionary REFERENCE=../KPNIH1.fasta OUTPUT=../KPNIH1.dict

</pre>
<blockquote>
<div><strong><em>ii. Run PICARD for removing duplicates.</em></strong></div></blockquote>
<pre class="literal-block">

java -jar /scratch/micro612w18_fluxod/shared/bin/picard-tools-1.130/picard.jar MarkDuplicates REMOVE_DUPLICATES=true INPUT=Rush_KPC_266__aln_sort.bam OUTPUT=Rush_KPC_266__aln_marked.bam METRICS_FILE=Rush_KPC_266__markduplicates_metrics CREATE_INDEX=true VALIDATION_STRINGENCY=LENIENT

</pre>
<p>The output of Picard remove duplicate step is a new bam file &#8220;Rush_KPC_266__aln_marked.bam&#8221; without PCR duplicates.</p>
<p>You will need to index this new marked.bam file for further processing.</p>
<blockquote>
<div><strong><em>iii. Index these marked bam file again using SAMTOOLS(For input in Artemis later)</em></strong></div></blockquote>
<pre class="literal-block">
samtools index Rush_KPC_266__aln_marked.bam
</pre>
<p>Open the markduplicates metrics file and glance through the number and percentage of PCR duplicates removed.
For more details about each metrics in a metrics file, please refer to <a class="reference external" href="https://broadinstitute.github.io/picard/picard-metric-definitions.html#DuplicationMetrics">this</a></p>
<pre class="literal-block">
nano Rush_KPC_266__markduplicates_metrics

#or 

less Rush_KPC_266__markduplicates_metrics
</pre>
</div>
<div class="section" id="generate-alignment-statistics">
<span id="generate-alignment-statistics"></span><h2>Generate Alignment Statistics<a class="headerlink" href="#generate-alignment-statistics" title="Permalink to this headline">¶</a></h2>
<p>Often, while analyzing sequencing data, we are required to make sure that our analysis steps are correct. Some statistics about our analysis will help us in making that decision. So Lets try to get some statistics about various outputs that were created using the above steps and check if everything makes sense.</p>
<blockquote>
<div><strong><em>i. Collect Alignment statistics using Picard</em></strong></div></blockquote>
<p>Run the below command on your marked.bam file</p>
<pre class="literal-block">

java -jar /scratch/micro612w18_fluxod/shared/bin/picard-tools-1.130/picard.jar CollectAlignmentSummaryMetrics R=../KPNIH1.fasta I=Rush_KPC_266__aln_marked.bam O=AlignmentSummaryMetrics.txt

</pre>
<p>Open the file AlignmentSummaryMetrics.txt and explore various statistics. It will generate various statistics and the definition for each can be found <a class="reference external" href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#AlignmentSummaryMetrics">here</a></p>
<p>The file AlignmentSummaryMetrics.txt contains many columns and at times it becomes difficult to extract information from a particular column if we dont know the exact column number. Run the below unix gem to print column name with its number.</p>
<pre class="literal-block">
grep 'CATEGORY' AlignmentSummaryMetrics.txt | tr '\t' '\n' | cat --number
</pre>
<ul class="simple">
<li>Question: Extract alignment percentage from AlignmentSummaryMetrics file. (% of reads aligned to reference genome)</li>
</ul>
<!---
awk -F'\t' '{print $7}' AlignmentSummaryMetrics.txt
--><pre class="literal-block">
grep -v '#' AlignmentSummaryMetrics.txt | cut -f7
</pre>
<p>Try to explore other statistics and their definitions from Picard AlignmentSummaryMetrics <a class="reference external" href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#AlignmentSummaryMetrics">link</a></p>
<blockquote>
<div><strong><em>ii. Estimate read coverage/read depth using Picard</em></strong></div></blockquote>
<p>Read coverage/depth describes the average number of reads that align to, or &#8220;cover,&#8221; known reference bases. The sequencing depth is one of the most crucial issue in the design of next-generation sequencing experiments. This <a class="reference external" href="https://www.nature.com/articles/nrg3642">paper</a> review current guidelines and precedents on the issue of coverage, as well as their underlying considerations, for four major study designs, which include de novo genome sequencing, genome resequencing, transcriptome sequencing and genomic location analyses</p>
<p>After read mapping, it is important to make sure that the reference bases are represented by enough read depth before making any inferences such as variant calling.</p>
<pre class="literal-block">
java -jar /scratch/micro612w18_fluxod/shared/bin/picard-tools-1.130/picard.jar CollectWgsMetrics R=../KPNIH1.fasta I=Rush_KPC_266__aln_marked.bam O=WgsMetrics.txt

</pre>
<p>Open the file &#8220;WgsMetrics.txt&#8221; and explore various statistics. It will generate various statistics and the definition for each can be found <a class="reference external" href="https://broadinstitute.github.io/picard/picard-metric-definitions.html#CollectWgsMetrics.WgsMetrics">here</a>.</p>
<p>Print column names</p>
<pre class="literal-block">
grep 'GENOME_TERRITORY' WgsMetrics.txt | tr '\t' '\n' | cat --number
</pre>
<p>Since &#8220;WgsMetrics.txt&#8221; also contains histogram information, we will run commands on only the first few lines to extract information.</p>
<ul class="simple">
<li>Question: Extract mean coverage information from &#8220;WgsMetrics.txt&#8221;</li>
</ul>
<!---
sed -n 7,8p WgsMetrics.txt | awk -F'\t' '{print $2}'
--><pre class="literal-block">
grep -v '#' WgsMetrics.txt | cut -f2 | head -n3

</pre>
<blockquote>
<div>Question: Percentage of bases that attained at least 5X sequence coverage.</div></blockquote>
<pre class="literal-block">
grep -v '#' WgsMetrics.txt | cut -f13 | head -n3
</pre>
<blockquote>
<div>Question: Percentage of bases that had siginificantly high coverage. Regions with unusually high depth sometimes indicate either repetitive regions or PCR amplification bias.</div></blockquote>
<pre class="literal-block">
grep -v '#' WgsMetrics.txt | cut -f25 | head -n3
</pre>
<!--
## Generate Alignment Statistics report using [Qualimap](http://qualimap.bioinfo.cipf.es/)
Qualimap outputs a very informative report about the alignments and coverage across the entire genome. Lets create one for our sample. The below command calls bamqc utility of qualimap and generates a report in pdf format.
``` 
qualimap bamqc -bam Rush_KPC_266__aln_sort.bam -outdir ./ -outfile Rush_KPC_266__report.pdf -outformat pdf 
```
Lets get this pdf report onto our local system and check the chromosome stats table, mapping quality and coverage across the entire reference genome.
```
scp username@flux-xfer.arc-ts.umich.edu:/scratch/micro612w18_fluxod/username/day1_after/Rush_KPC_266_varcall_result/Rush_KPC_266__report.pdf /path-to-local-directory/
```
--></div>
<div class="section" id="variant-calling-and-filteration">
<span id="variant-calling-and-filteration"></span><h2>Variant Calling and Filteration<a class="headerlink" href="#variant-calling-and-filteration" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="day1_afternoon.html">[back to top]</a>
<a class="reference external" href="index.html">[HOME]</a></p>
<p>One of the downstream uses of read mapping is finding differences between our sequence data against a reference. This step is achieved by carrying out variant calling using any of the variant callers (samtools, gatk, freebayes etc). Each variant caller uses a different statistical framework to discover SNPs and other types of mutations. For those of you who are interested in finding out more about the statistics involved, please refer to <a class="reference external" href="">this</a> samtools paper, one of most commonly used variant callers.</p>
<p>The <a class="reference external" href="https://www.broadinstitute.org/gatk/guide/best-practices.php">GATK best practices guide</a> will provide more details about various steps that you can incorporate in your analysis.</p>
<p>There are many published articles that compare different variant callers but this is a very interesting <a class="reference external" href="https://bcbio.wordpress.com/2013/10/21/updated-comparison-of-variant-detection-methods-ensemble-freebayes-and-minimal-bam-preparation-pipelines/">blog post</a> that compares the performance and accuracy of different variant callers.</p>
<p>Here we will use samtools mpileup to perform this operation on our BAM file and generate a VCF (variant call format) file.</p>
<p><strong>1. Call variants using <a class="reference external" href="http://www.htslib.org/doc/samtools.html">samtools</a> mpileup and <a class="reference external" href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a></strong></p>
<pre class="literal-block">

samtools mpileup -ug -f ../KPNIH1.fasta Rush_KPC_266__aln_marked.bam | bcftools call -O v -v -c -o Rush_KPC_266__aln_mpileup_raw.vcf


#In the above command, we are using samtools mpileup to generate a pileup formatted file from BAM alignments and genotype likelihoods (-g flag) in BCF format (binary version of vcf). This bcf output is then piped to bcftools, which calls variants and outputs them in vcf format (-c flag for using consensus calling algorithm  and -v for outputting variants positions only)


</pre>
<p>Let&#8217;s go through the VCF file and try to understand a few important VCF specifications and criteria that we can use for filtering low confidence SNPs.</p>
<pre class="literal-block">
less Rush_KPC_266__aln_mpileup_raw.vcf
</pre>
<ol class="simple">
<li>CHROM, POS: 1st and 2nd column represent the reference genome name and reference base position where a variant was called</li>
<li>REF, ALT: 4th and 5th columns represent the reference allele at the position and alternate/variant allele called from the reads</li>
<li>QUAL: Phred-scaled quality score for the assertion made in ALT</li>
<li>INFO: Additional information that provides technical scores and obervations for each variant. Important parameters to look for: Depth (DP), mapping quality (MQ), FQ (consensus score), allele frequency for each ALT allele (AF)</li>
</ol>
<p>VCF format stores a large variety of information and you can find more details in <a class="reference external" href="https://www.google.com/url?sa=t&amp;rct=j&amp;q=&amp;esrc=s&amp;source=web&amp;cd=1&amp;ved=0ahUKEwit35bvktzLAhVHkoMKHe3hAhYQFggdMAA&amp;url=https%3A%2F%2Fsamtools.github.io%2Fhts-specs%2FVCFv4.2.pdf&amp;usg=AFQjCNGFka33WgRmvOfOfp4nSaCzkV95HA&amp;sig2=tPLD6jW5ALombN3ALRiCZg&amp;cad=rja">this pdf</a>.</p>
<p>Lets count the number of raw unfiltered variants found:</p>
<pre class="literal-block">
grep -v '#' Rush_KPC_266__aln_mpileup_raw.vcf | wc -l

grep -v '#' Rush_KPC_266__aln_mpileup_raw.vcf | grep 'INDEL' | wc -l
</pre>
<p><strong>2. Variant filtering and processed file generation using GATK and vcftools</strong></p>
<blockquote>
<div><strong><em>i. Variant filtering using <a class="reference external" href="https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_filters_VariantFiltration.php">GATK</a>:</em></strong></div></blockquote>
<p>There are various tools that can you can try for variant filteration such as vcftools, GATK, vcfutils etc. Here we will use GATK VariantFiltration utility to filter out low confidence variants.</p>
<p>Run this command on raw vcf file Rush_KPC_266__aln_mpileup_raw.vcf.</p>
<pre class="literal-block">

java -jar /scratch/micro612w18_fluxod/shared/bin/GenomeAnalysisTK-3.3-0/GenomeAnalysisTK.jar -T VariantFiltration -R ../KPNIH1.fasta -o Rush_KPC_266__filter_gatk.vcf --variant Rush_KPC_266__aln_mpileup_raw.vcf --filterExpression &quot;FQ &lt; 0.025 &amp;&amp; MQ &gt; 50 &amp;&amp; QUAL &gt; 100 &amp;&amp; DP &gt; 15&quot; --filterName pass_filter

</pre>
<p>This command will add a &#8216;pass_filter&#8217; text in the 7th FILTER column for those variant positions that passed our filtered criteria:</p>
<ol class="simple">
<li>DP: Depth of reads. More than 15 reads supporting a variant call at these position.</li>
<li>MQ: Root Mean Square Mapping Quality. This provides an estimation of the overall mapping quality of reads supporting a variant call. The root mean square is equivalent to the mean of the mapping qualities plus the standard deviation of the mapping qualities.</li>
<li>QUAL stands for phred-scaled quality score for the assertion made in ALT. High QUAL scores indicate high confidence calls.</li>
<li>FQ stands for consensus quality. A positive value indicates heterozygote and a negative value indicates homozygous. In bacterial analysis, this plays an important role in defining if a gene was duplicated in a particular sample. We will learn more about this later while visualizing our BAM files in Artemis.</li>
</ol>
<p>Check if the pass_filter was added properly and count the number of variants that passed the filter.</p>
<pre class="literal-block">
grep 'pass_filter' Rush_KPC_266__filter_gatk.vcf | head

</pre>
<p><strong><em>Caveat: This filter criteria should be applied carefully after giving some thought to the type of library, coverage, average mapping quality, type of analysis and other such requirements.</em></strong></p>
<blockquote>
<div><strong><em>ii. Remove indels and keep only SNPS that passed our filter criteria using <a class="reference external" href="http://vcftools.sourceforge.net/man_latest.html">the vcftools manual</a>:</em></strong></div></blockquote>
<p>vcftools is a program package that is especially written to work with vcf file formats. It thus saves your precious time by making available all the common operations that you would like to perform on the vcf file using a single command. One such operation is removing INDEL information from a vcf file.</p>
<p>Now, let&#8217;s remove indels from our final vcf file and keep only variants that passed our filter criteria (positions with pass_filter in their FILTER column).</p>
<pre class="literal-block">

vcftools --vcf Rush_KPC_266__filter_gatk.vcf --keep-filtered pass_filter --remove-indels --recode --recode-INFO-all --out Rush_KPC_266__filter_onlysnp

</pre>
<!--
commenting out consensus generation
>iii. Generate Consensus fasta file from filtered variants using vcftools:
A consensus fasta sequence will contain alleles from reference sequence at positions where no variants were observed and variants that were observed at positions described in vcf file.
Run the commands below to generate a consensus fasta sequence.
```
bgzip Rush_KPC_266__filter_onlysnp.recode.vcf
tabix Rush_KPC_266__filter_onlysnp.recode.vcf.gz
cat ../KPNIH1.fasta | vcf-consensus Rush_KPC_266__filter_onlysnp.recode.vcf.gz > Rush_KPC_266__consensus.fa
```
> Note: Dont forget to put the actual path to the refeerence sequence in place of /path-to-reference/
Check the fasta header and change it using sed.
```
head -n1 Rush_KPC_266__consensus.fa
sed -i 's/>.*/>Rush_KPC_266_/g' Rush_KPC_266__consensus.fa 
```
--><p><strong><em>3. Variant Annotation using snpEff</em></strong></p>
<p>Variant annotation is one of the crucial steps in any variant calling pipeline. Most of the variant annotation tools create their own database or use an external one to assign function and predict the effect of variants on genes. We will try to touch base on some basic steps of annotating variants in our vcf file using snpEff.</p>
<p>You can annotate these variants before performing any filtering steps that we did earlier or you can decide to annotate just the final filtered variants.</p>
<p>snpEff contains a database of about 20,000 reference genomes built from trusted and public sources. Lets check if snpEff contains a database of our reference genome.</p>
<blockquote>
<div><strong><em>i. Check snpEff internal database for your reference genome:</em></strong></div></blockquote>
<div class="highlight-"><div class="highlight"><pre>java -jar /scratch/micro612w18_fluxod/shared/bin/snpEff/snpEff.jar databases | grep &#39;kpnih1&#39;
</pre></div>
</div>
<p>Note down the genome id for your reference genome KPNIH1. In this case: GCA_000281535.2.29</p>
<blockquote>
<div><strong><em>ii. Change the chromosome name in the vcf file to ‘Chromosome’ for snpEff reference database compatibility.</em></strong></div></blockquote>
<pre class="literal-block">
sed -i 's/gi.*|/Chromosome/g' Rush_KPC_266__filter_gatk.vcf
</pre>
<blockquote>
<div><strong><em>iii. Run snpEff for variant annotation.</em></strong></div></blockquote>
<pre class="literal-block">

java -jar /scratch/micro612w18_fluxod/shared/bin/snpEff/snpEff.jar -onlyProtein -no-upstream -no-downstream  -no-intergenic -v GCA_000281535.2.29 Rush_KPC_266__filter_gatk.vcf &gt; Rush_KPC_266__filter_gatk_ann.vcf -csvStats Rush_KPC_266__filter_gatk_stats

</pre>
<p>The STDOUT  will print out some useful details such as genome name and version being used, no. of genes, protein-coding genes and transcripts, chromosome and plasmid names etc.</p>
<p>snpEff will add an extra field named &#8216;ANN&#8217; at the end of INFO field. Lets go through the ANN field added after annotation step.</p>
<pre class="literal-block">
grep 'ANN=' Rush_KPC_266__filter_gatk_ann.vcf | head -n1

or to print on seperate lines

grep -o 'ANN=.*GT:PL' Rush_KPC_266__filter_gatk_ann.vcf | head -n1 | tr '|' '\n' | cat --number
</pre>
<p>The ANN field will provide information such as the impact of variants (HIGH/LOW/MODERATE/MODIFIER) on genes and transcripts along with other useful annotations.</p>
<p>Detailed information of the ANN field and sequence ontology terms that it uses can be found <a class="reference external" href="http://snpeff.sourceforge.net/SnpEff_manual.html#input">here</a>.</p>
<p>Let&#8217;s see how many SNPs and Indels passed the filter using grep and wc.</p>
<pre class="literal-block">

No. of Variants:
grep '^Chromosome' Rush_KPC_266__filter_gatk_ann.vcf | wc -l

No. of Variants that passed the filter:
grep '^Chromosome.*pass_filter' Rush_KPC_266__filter_gatk_ann.vcf | wc -l

No. of SNPs that passed the filter:
grep '^Chromosome.*pass_filter' Rush_KPC_266__filter_gatk_ann.vcf | grep -v 'INDEL' | wc -l

No. of Indels that passed the filter:
grep '^Chromosome.*pass_filter' Rush_KPC_266__filter_gatk_ann.vcf | grep 'INDEL' | wc -l


</pre>
</div>
<div class="section" id="visualize-bam-and-vcf-files-in-artemis">
<span id="visualize-bam-and-vcf-files-in-artemis"></span><h2>Visualize BAM and VCF files in <a class="reference external" href="http://www.sanger.ac.uk/science/tools/artemis">Artemis</a><a class="headerlink" href="#visualize-bam-and-vcf-files-in-artemis" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="day1_afternoon.html">[back to top]</a>
<a class="reference external" href="index.html">[HOME]</a></p>
<p>While these various statistical/text analyses are helpful, visualization of all of these various output files can help in making some significant decisions and inferences about your entire analysis. There are a wide variety of visualization tools out there that you can choose from for this purpose.</p>
<p>We will be using <a class="reference external" href="http://www.sanger.ac.uk/science/tools/artemis">Artemis</a> here, developed by the Sanger Institute for viewing BAM and vcf files for manual inspection of some of the variants.</p>
<ul class="simple">
<li><strong><em>Required Input files:</em></strong></li>
</ul>
<blockquote>
<div>KPNIH1 reference fasta
KPNIH1 genbank file
Rush_KPC_266__aln_marked.bam
Rush_KPC_266__aln_marked.bam.bai
Rush_KPC_266__filter_gatk_ann.vcf.gz
Rush_KPC_266__filter_gatk_ann.vcf.gz.tbi</div></blockquote>
<p>Let&#8217;s make a seperate folder (make sure you are in the Rush_KPC_266_varcall_result folder) for the files that we need for visualization and copy it to that folder</p>
<pre class="literal-block">

mkdir Artemis_files

cp ../KPNIH1.fasta ../KPNIH.gb Rush_KPC_266__aln_marked.bam Rush_KPC_266__aln_marked.bam.bai Rush_KPC_266__filter_gatk_ann.vcf Artemis_files/

</pre>
<p>We need to replace the genome name that we changed earlier for snpEff. (Make sure you are in Artemis_files folder)</p>
<pre class="literal-block">

cd Artemis_files

sed -i 's/Chromosome/gi|661922017|gb|CP008827.1|/g' Rush_KPC_266__filter_gatk_ann.vcf 

bgzip Rush_KPC_266__filter_gatk_ann.vcf

tabix Rush_KPC_266__filter_gatk_ann.vcf.gz
</pre>
<p>Open a new terminal and run the scp command or cyberduck to get these files to your local system.</p>
<pre class="literal-block">

scp -r username&#64;flux-xfer.arc-ts.umich.edu:/scratch/micro612w18_fluxod/username/day1_after/Rush_KPC_266_varcall_result/Artemis_files/ /path-to-local-directory/

#You can use ~/Desktop/ as your local directory path
</pre>
<p>Start Artemis.</p>
<p>Set your working directory to Artemis_files (the Artemis_files folder that you copied to your local system) by clicking the browse button  and click OK.</p>
<p>Now go to the top left File options and select Open File Manager. You should see the folder Artemis_files. Expand it and select KPNIH.gb file. A new window should open displaying your features stored in a genbank file.</p>
<p>Now open the BAM file by selecting File (Top left corner) -&gt; Read BAM/VCF file -&gt; Select -&gt; Rush_KPC_266__aln_marked.bam -&gt; OK</p>
<p>Reads aligned to your reference are displayed as stacked at the top panel of Artemis. The reads are color-coded so that paired reads are blue and those with an inversion are red. Reads that do not have a mapped mate are black and are optionally shown in the inferred insert size view. In the stack view, duplicated reads that span the same region are collapsed into one green line.</p>
<p>Now right click on any of the stacked reads and Go to Graph and select Coverage (screenshot below).</p>
<p>Now right click on any of the stacked reads and Go to Show and select SNP marks to show SNPs in red marks.</p>
<p><img alt="alt tag" src="_images/select_graph.png" /></p>
<p>Follow the same procedure and select SNP graph. Adjust the gene features panel height to show all the graph in a window.</p>
<p><img alt="alt tag" src="_images/graphs.png" /></p>
<p>Play around by moving the genbank panel cursor to look at coverage and SNP density across the genome. This will let you look at any regions where the coverage or SNP density is unusually high or low.</p>
<p>If you click a read, its mate pair will also be selected. If the cursor hovers over a read for long enough details of that read will appear in a small box. For more details of the read, right-click and select &#8216;Show details of: READ NAME&#8217; (last option in list) from the
menu (screenshot below). This will open up a new window giving you some useful details such as mapping quality, coordinates etc.</p>
<p><img alt="alt tag" src="_images/read_details.png" /></p>
<p>The snps are denoted by red marks as observed inside the reads. Go to one of the SNPs in the VCF file (Position: 50195) by directly navigating to the position. For this, select Goto at the top -&gt; select Navigator -&gt; Type the position in Goto Base box</p>
<p>You will notice a spike in the middle of the SNP graph window. This is one of the SNPs that passed all our filter criteria. (Screenshot)</p>
<p><img alt="alt tag" src="_images/spike_true.png" /></p>
<p>Lets try to see an example of HET variant. Variant positions where more than one allele (variant) with sufficiently high read depth are observed are considered HET type variants.</p>
<p>For this, click on tje Goto option at the top and select navigator. Type 321818 in Goto Base box and click Goto.</p>
<p>You will see a thick spike in the SNP graph as well as thick red vertical line in BAM panel. Also notice the sudden spike in the coverage for this particular region compared to its flanking region (the region before and after a selected region). The coverage here is more than 300 which is unusually high compared to the entire genome coverage. This means that more than one allele with high quality and depth were observed at these positions so we cannot decide which one of these is a true variant. We removed these types of variants during our Variant Filteration step using the criteria FQ. (If the FQ is unusually high, it is suggestive of a HET variant and negative FQ value is a suggestive of true variant as observed in the mapped reads)</p>
<p><img alt="alt tag" src="_images/HET_variant.png" /></p>
<p>Now select the gene right below this spiked region. Right click on this gene (KPNIH1_RS01560) and select Zoom to Selection.</p>
<p><img alt="alt tag" src="_images/HET_variant_gene_selected.png" /></p>
<p>Check the details about gene by selecting View -&gt; Selected Features</p>
<p>You can inspect these type of HET variants later for any gene duplication or copy number analysis (by extracting variant positions with high FQ values). Addition of these details will give a better resolution while inferring phylogenetic trees.</p>
<p>Play around with Artemis to look at what other kind of information you can find from these BAM and vcf files. Also refer to the manual at the <a class="reference external" href="http://www.sanger.ac.uk/science/tools/artemis">Artemis Homepage</a> for full information about its usage.</p>
<p><a class="reference external" href="day1_afternoon.html">[back to top]</a>
<a class="reference external" href="index.html">[HOME]</a></p>
</div>
<div class="section" id="vre-variant-calling-analysis">
<span id="vre-variant-calling-analysis"></span><h2>VRE variant calling analysis<a class="headerlink" href="#vre-variant-calling-analysis" title="Permalink to this headline">¶</a></h2>
<p>Today, we learned how to assess the quality, perform quality trimming and variant calling to find variants between the sample and reference genome. This exercise requires you to apply these tools and commands on a new data set. These samples both come from a patient infected with VRE before and after treatment with daptomycin. The first sample was the patients initial sample and is susceptible to daptomycin, and the second was after daptomycin resistance emerged during treatment. Your goal is to map reads from the resistant genome to the susceptible reference and search for variants that may be associated with resistance. To accomplish this you will run the programs from this session to generate filtered variant files (VCF), and then explore these variants in Artemis to see what genes they are in. To help with your interpretation, see if you see any genes hit that were reported in this <a class="reference external" href="http://www.nejm.org/doi/full/10.1056/nejmoa1011138">paper</a>, which was the first to idenitfy putative daptomycin resistance loci.</p>
<ul class="simple">
<li>Use VRE_daptoS_ref_strain.fa as your reference genome and VRE_daptoS_gene_annot.gff annotation file for Artemis.</li>
<li>This is how the command and tools workflow should look like:</li>
</ul>
<blockquote>
<div><ol class="simple">
<li>FastQC to check the quality of reads(you can skip here for time)</li>
<li>Trimmomatic to remove bad quality data(you can skip here for time)</li>
<li>Prepare reference genome index for BWA and align reads to reference genome</li>
<li>SAM/BAM manipulation using samtools</li>
<li>Remove duplicates using picard(dont forget to create a dictionary for reference fasta file required by PICARD)</li>
<li>Index marked bam file generated by picard using SAMTOOLS(For input in Artemis later)</li>
<li>Variant calling using samtools</li>
<li>Variant Filteration using GATK</li>
<li>Visualize BAM and VCF files in Artemis</li>
</ol>
</div></blockquote>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="day2_morning.html" class="btn btn-neutral float-right" title="Day 2 Morning">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="day1_morning.html" class="btn btn-neutral" title="Day 1 Morning"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Evan Snitkin, Ali Pirani.
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
  
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'3.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>