

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Helpful resources for microbial genomics &mdash; Micro 612 genomics workshop 3.0 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700&subset=latin,cyrillic' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Micro 612 genomics workshop 3.0 documentation" href="index.html"/>
        <link rel="prev" title="Day 3 Afternoon" href="day3_afternoon.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        
          <a href="index.html" class="fa fa-home"> Micro 612 genomics workshop</a>
        
        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="day1_morning.html">Day 1 Morning</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#installing-and-setting-up-cyberduck-for-file-transfer">Installing and setting up Cyberduck for file transfer</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#getting-your-data-onto-flux-and-setting-up-environment-variable">Getting your data onto Flux and setting up environment variable</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#unix-is-your-friend">Unix is your friend</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#your-first-sequence-analysis-program">Your first sequence analysis program!!!</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#plotting-genomic-coverage-in-r">Plotting genomic coverage in R</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#power-of-unix-commands">Power of Unix commands</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#contamination-screening-using-fastq-screen">Contamination Screening using FastQ Screen</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#quality-control-using-fastqc">Quality Control using FastQC</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_morning.html#quality-trimming-using-trimmomatic">Quality Trimming using Trimmomatic</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="day1_afternoon.html">Day 1 Afternoon</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day1_afternoon.html#read-mapping">Read Mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_afternoon.html#generate-alignment-statistics">Generate Alignment Statistics</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_afternoon.html#variant-calling-and-filteration">Variant Calling and Filteration</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_afternoon.html#visualize-bam-and-vcf-files-in-artemis">Visualize BAM and VCF files in Artemis</a></li>
<li class="toctree-l2"><a class="reference internal" href="day1_afternoon.html#vre-variant-calling-analysis">VRE variant calling analysis</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="day2_morning.html">Day 2 Morning</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#genome-assembly-using-spades-pipeline">Genome Assembly using Spades Pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#assembly-evaluation-using-quast">Assembly evaluation using QUAST</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#generating-multiple-sample-reports-using-multiqc">Generating multiple sample reports using multiqc</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#compare-assembly-to-reference-genome-and-post-assembly-genome-improvement">Compare assembly to reference genome and post-assembly genome improvement</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#map-reads-to-the-final-ordered-assembly">Map reads to the final ordered assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#using-abacas-and-act-to-compare-vre-vse-genome">Using abacas and ACT to compare VRE/VSE genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_morning.html#genome-annotation">Genome Annotation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="day2_afternoon.html">Day 2 Afternoon</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#high-throughput-blast-and-pan-genome-analysis">High-throughput BLAST and pan-genome analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#determine-which-genomes-contain-beta-lactamase-genes">Determine which genomes contain beta-lactamase genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#identification-of-antibiotic-resistance-genes-with-ariba-directly-from-paired-end-reads">Identification of antibiotic resistance genes with ARIBA directly from paired end reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#perform-pan-genome-analysis-with-roary">Perform pan-genome analysis with Roary</a></li>
<li class="toctree-l2"><a class="reference internal" href="day2_afternoon.html#perform-genome-comparisons-with-act">Perform genome comparisons with ACT</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="day3_morning.html">Day 3 Morning</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#perform-whole-genome-alignment-with-mauve-and-convert-alignment-to-other-useful-formats">Perform whole genome alignment with Mauve and convert alignment to other useful formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#perform-some-dna-sequence-comparisons-and-phylogenetic-analysis-in-ape-an-r-package">Perform some DNA sequence comparisons and phylogenetic analysis in APE, an R package</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#perform-snp-density-analysis-to-discern-evidence-of-recombination">Perform SNP density analysis to discern evidence of recombination</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#perform-recombination-filtering-with-gubbins">Perform recombination filtering with Gubbins</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_morning.html#create-annotated-publication-quality-trees-with-itol">Create annotated publication quality trees with iTOL</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="day3_afternoon.html">Day 3 Afternoon</a><ul>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#klebsiella-pneumoniae-comparative-genomic-analysis">Klebsiella pneumoniae comparative genomic analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#perform-qc-on-fastq-files">Perform QC on fastq files</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#examine-results-of-spandx-pipeline">Examine results of SPANDx pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#recombination-detection-and-tree-generation">Recombination detection and tree generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#phylogenetic-tree-annotation-and-visualization">Phylogenetic tree annotation and visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="day3_afternoon.html#assessment-of-genomic-deletions">Assessment of genomic deletions</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Helpful resources for microbial genomics</a></li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Micro 612 genomics workshop</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Helpful resources for microbial genomics</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/online_resources.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="helpful-resources-for-microbial-genomics">
<span id="helpful-resources-for-microbial-genomics"></span><h1>Helpful resources for microbial genomics<a class="headerlink" href="#helpful-resources-for-microbial-genomics" title="Permalink to this headline">¶</a></h1>
<p><strong><em>If you were not able to follow the video, here is the <a class="reference external" href="https://www.youtube.com/watch?v=womKfikWlxM">link</a> to illumina Sequencing</em></strong></p>
<p><a class="reference external" href="index.html">[HOME]</a></p>
<p><strong>General Bioinformatics resources</strong></p>
<ul class="simple">
<li><a class="reference external" href="http://omictools.com/">Omictools</a></li>
<li><a class="reference external" href="https://github.com/stephenturner/oneliners">Bioinformatics One-liners by Stephen Turner</a></li>
<li><a class="reference external" href="https://sequencing.qcfail.com/">QC Fail: Explaining your errors</a></li>
</ul>
<p><strong>Short read processing</strong></p>
<ul class="simple">
<li><a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></li>
<li>Trimmomatic: <a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">Home</a> <a class="reference external" href="http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf">Manual</a></li>
<li><a class="reference external" href="http://bio-bwa.sourceforge.net/">bwa</a></li>
<li><a class="reference external" href="http://bowtie-bio.sourceforge.net/index.shtml">bowtie</a></li>
<li><a class="reference external" href="http://samtools.sourceforge.net/">samtools</a></li>
<li><a class="reference external" href="http://vcftools.sourceforge.net/">vcftools</a></li>
<li><a class="reference external" href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a></li>
<li><a class="reference external" href="https://www.broadinstitute.org/gatk/">gatk</a></li>
<li><a class="reference external" href="http://broadinstitute.github.io/picard/">picard</a></li>
<li><a class="reference external" href="https://github.com/dsarov/SPANDx">SPANDx</a></li>
<li><a class="reference external" href="https://github.com/tseemann/snippy">Snippy</a></li>
</ul>
<p><strong>Genome assembly</strong></p>
<ul class="simple">
<li><a class="reference external" href="http://bioinf.spbau.ru/spades">Spades</a></li>
<li><a class="reference external" href="https://www.ebi.ac.uk/~zerbino/velvet/">Velvet</a></li>
<li><a class="reference external" href="https://sourceforge.net/p/mira-assembler/wiki/Home/">Mira</a></li>
<li><a class="reference external" href="https://sourceforge.net/p/ngopt/wiki/browse_pages/">A5</a></li>
</ul>
<p><strong>Genome alignment</strong></p>
<ul class="simple">
<li><a class="reference external" href="http://darlinglab.org/mauve/download.html">Mauve</a></li>
<li><a class="reference external" href="http://mummer.sourceforge.net/">MUMmer</a></li>
<li><a class="reference external" href="http://mugsy.sourceforge.net/">Mugsy</a></li>
</ul>
<p><strong>Visualization of genomic data</strong></p>
<ul class="simple">
<li><a class="reference external" href="http://www.sanger.ac.uk/science/tools/artemis">Artemis</a></li>
<li><a class="reference external" href="http://www.sanger.ac.uk/science/tools/artemis-comparison-tool-act">Artemis Comparison Tool</a></li>
<li><a class="reference external" href="https://www.broadinstitute.org/igv/">IGV</a></li>
</ul>
<p><strong>Genome annotation</strong></p>
<ul class="simple">
<li><a class="reference external" href="http://www.vicbioinformatics.com/software.prokka.shtml">Prokka</a></li>
<li><a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">Blastall</a></li>
<li><a class="reference external" href="https://github.com/jasonsahl/LS-BSR">LS-BSR</a></li>
</ul>
<p><strong>Phylogenetic tools and resources</strong></p>
<ul class="simple">
<li>Visualization<ul>
<li><a class="reference external" href="http://doua.prabi.fr/software/seaview">Seaview</a></li>
<li><a class="reference external" href="http://itol.embl.de/">iTOL</a></li>
<li><a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">Figtree</a></li>
</ul>
</li>
<li>Phylogenetic software<ul>
<li><a class="reference external" href="http://paup.csit.fsu.edu/">PAUP</a></li>
<li><a class="reference external" href="http://sco.h-its.org/exelixis/software.html">RaxML</a></li>
<li><a class="reference external" href="http://www.atgc-montpellier.fr/phyml/">PhyML</a></li>
<li><a class="reference external" href="http://beast.bio.ed.ac.uk/">BEAST</a></li>
<li><a class="reference external" href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP</a></li>
<li><a class="reference external" href="http://ape-package.ird.fr/">APE</a></li>
<li><a class="reference external" href="http://microreact.org/showcase/">Microreact</a></li>
<li>Recombination detection<ul>
<li><a class="reference external" href="http://www.sanger.ac.uk/science/tools/gubbins">Gubbins</a></li>
<li><a class="reference external" href="http://www.xavierdidelot.xtreemhost.com/clonalframe.htm?ckattempt=1">ClonalFrame</a></li>
<li><a class="reference external" href="http://evolution.genetics.washington.edu/phylip/software.html">List of Phylogeny Programs</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p><strong>Databases</strong></p>
<ul class="simple">
<li><a class="reference external" href="http://ardb.cbcb.umd.edu/">ARDB</a></li>
<li><a class="reference external" href="https://www.patricbrc.org/portal/portal/patric/Home">PATRIC</a></li>
</ul>
<p><strong>Video Resources you should watch and follow</strong></p>
<ul class="simple">
<li><a class="reference external" href="https://www.youtube.com/watch?v=bz93ReOv87Y">FastQC</a></li>
<li><a class="reference external" href="https://www.youtube.com/user/GenomeTV">NHGRI</a></li>
<li><a class="reference external" href="https://www.youtube.com/channel/UCv4IbnP9j9RC_aZAs8wqdeQ">Broad Institute</a></li>
<li><a class="reference external" href="https://www.youtube.com/channel/UCVqWctrxf5-oBIM1lqOIt-A">Cold Spring Harbor Lab</a></li>
<li><a class="reference external" href="https://www.youtube.com/user/NCBINLM/videos">NCBI</a></li>
<li><a class="reference external" href="https://www.youtube.com/channel/UCEBb1b_L6zDS3xTUrIALZOw">Bioinformatics courses from MIT</a></li>
<li>Youtube <a class="reference external" href="https://www.youtube.com/channel/UC1lb9cYp9wt8xjF3APM9bMw">Channel</a> of <a class="reference external" href="http://rafalab.dfci.harvard.edu/">Rafael Irizarry</a> covering various topics on NGS analysis and statistics involved in it.</li>
</ul>
<p><strong><a class="reference external" href="http://www.acgt.me/blog/2014/3/25/101-questions-a-new-series-of-interviews-with-notable-bioinformaticians">101 Questions: a series of interviews with notable bioinformaticians</a></strong></p>
<p><strong><a class="reference external" href="http://cabbagesofdoom.blogspot.com/2015/08/bioinformatics-is-just-like-bench.html">Bioinformatics is just like bench science and should be treated as such</a></strong></p>
<p><strong>Unix/Command line</strong></p>
<ul class="simple">
<li><a class="reference external" href="http://rik.smith-unna.com/command_line_bootcamp/?id=9xnbkx6eaof">command-line bootcamp</a></li>
<li><a class="reference external" href="https://www.codecademy.com/en/courses/learn-the-command-line">Code academy</a></li>
</ul>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="day3_afternoon.html" class="btn btn-neutral" title="Day 3 Afternoon"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Evan Snitkin, Ali Pirani.
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
  
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'3.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>